cogs/utils/music_helpers.py

```diff
@@ -0,0 +1,116 @@
+import time
+from functools import wraps
+from typing import Iterable
+import youtube_dl
+import discord
+import random
+
+FFMPEG_OPTS = "-nostats -loglevel quiet -vn"
+FFMPEG_BEFORE_OPTS = "-reconnect 1 -reconnect_streamed 1 -reconnect_delay_max 5"
+
+YTDL = youtube_dl.YoutubeDL({
+    "format": "bestaudio/best",
+    "restrictfilenames": True,
+    "noplaylist": True,
+    "quiet": True,
+    "no_warnings": True,
+    "default_search": "auto",
+    "geo_bypass": True,
+})
+
+QUEUE_ACTIONS = {
+    "⏪": lambda i, _l: (0, i != 0),
+    "◀": lambda i, l: ((i - 1) % l, True),
+    "▶": lambda i, l: ((i + 1) % l, True),
+    "⏩": lambda i, l: (l - 1, i != l - 1),
+}
+
+
+def check_playing(func):
+    @wraps(func)
+    async def wrapper(self, ctx, *args, **kwargs):
+        if (self.playing is None or ctx.voice_client is None
+                or (not self.track_lock.locked())):
+            await ctx.send(
+                "bot is not currently playing anything to a voice channel.")
+        elif (ctx.author.voice is None
+              or ctx.author.voice.channel != ctx.voice_client.channel) and len(
+                  ctx.voice_client.channel.members) > 1:
+            await ctx.send(
+                "you must be listening to music with the bot do this.")
+        else:
+            await func(self, ctx, *args, **kwargs)
+
+    return wrapper
+
+
+def mk_title_string(inf_dict) -> str:
+    url = inf_dict.get("webpage_url")
+    return (inf_dict.get("title", "title not found") if url is None else
+            f"[{inf_dict.get('title', 'title not found')}]({url})")
+
+
+def mk_duration_string(track_iter: Iterable) -> str:
+    total: int = 0
+    for track in track_iter:
+        if isinstance(track, tuple):
+            track = track[0]
+        dur = track.get("duration")
+        if dur is None:
+            return "n/a (livestream)"
+        total += dur
+    return time.strftime("%H:%M:%S", time.gmtime(total)) if (
+        total > 0 or not list(track_iter)) else "n/a (livestream)"
+
+
+def parse_time(time_str: str) -> int:
+    total: int = 0
+    for index, substr in enumerate(reversed(time_str.split(":"))):
+        if substr.isdigit():
+            total += int(substr) * (60**index)
+        else:
+            raise ValueError
+    return total
+
+
+def time_func(func):
+    async def wrapper(self, ctx, time_str: str):
+        try:
+            parsed = parse_time(time_str)
+        except ValueError:
+            await ctx.send(f"could not parse `{time_str}` to a time value.")
+            return
+        seconds = func(self, parsed, ctx.voice_client.is_paused())
+        self.skip_opts = time.strftime("%H:%M:%S",
+                                       time.gmtime(seconds)), seconds
+        self.track_start_time = time.perf_counter() - seconds
+        ctx.voice_client.stop()
+        await ctx.send(
+            f"moved to `{self.skip_opts[0]}` in currently playing track.")
+
+    wrapper.__name__ = func.__name__
```

Comment:
there is, its functools.wraps, but i don't use it here because it also overrides the function the wrapper type signature with the input function (which i don't wanna do here because i'm changing the output signature on purpose)